Feat: per chain safe mode to restrict deposits

msgmaxim opened the pull request:
# Pull Request

Closes: PRO-1218

## Checklist

Please conduct a thorough self-review before opening the PR.

- [x] I am confident that the code works.
- [ ] I have updated documentation where appropriate.

## Summary

Note that this is per chain rather than per asset (but per chain is easier to implement and after chatting with @kylezs it seems that per chain is what we actually want here).

- Added `deposits_enabled` to ingress-egress pallet's safe mode, which can be toggled to disable/enable openning of deposit channels for the target chain.
- Added a unit test to show that disabling `deposits_enabled` results in Errors on trying to open a channel.
- Anabled CODE_GREEN on ingress-egress pallets since we no longer want to use safe mode to prevent premature interactions with boost pools (see https://github.com/chainflip-io/chainflip-backend/pull/4816), and CODE_RED would now disable all deposits, which we don't want.

j4m1ef0rd commented on state-chain/pallets/cf-ingress-egress/src/tests/boost.rs:
```diff
@@ -795,31 +795,20 @@ fn test_add_boost_funds() {
+	let boost_mode = if enable { PalletSafeMode::CODE_GREEN } else { PalletSafeMode::CODE_RED };
+
+	let new_mode =
+		PalletSafeMode { deposits_enabled: get_safe_mode().deposits_enabled, ..boost_mode };
```
This is ok, but i think it would be better to just set the individual bool's in the test, like what you did in the `safe_mode_prevents_deposit`. That way we can make sure we are using the correct flag as well. To keep it clean we can still use the old `turn_safe_mode_off` function.

msgmaxim replied:
I think right now it is not clear which approach is better. When we have more bools in `PalletSafeMode`, we will have a better idea, and maybe we could revisit then.

kylezs commented on state-chain/pallets/cf-ingress-egress/src/lib.rs:
```diff
@@ -718,6 +721,8 @@ pub mod pallet {
 		/// Retrieving boost funds disabled due to safe mode.
 		StopBoostingDisabled,
+		/// Disabled due to safe mode for the chain
+		DepositsDisabledForChain,
```
DepositChannelCreationDisabled - technically deposits are still possible via already open channels or via contract initiated swaps

kylezs commented on state-chain/pallets/cf-ingress-egress/src/tests.rs:
```diff
@@ -1794,3 +1794,34 @@ fn ingress_fee_is_withheld_or_scheduled_for_swap() {
 }
+
+#[test]
+fn safe_mode_prevents_deposit() {
```
```suggestion
fn safe_mode_prevents_deposit_channel_creation() {
```

kylezs commented on state-chain/pallets/cf-ingress-egress/src/tests/boost.rs:
```diff
@@ -836,7 +825,7 @@ fn boosting_deposits_is_disabled_by_safe_mode() {
 		));
 
-		turn_safe_mode_on();
+		toggle_boosting_with_safe_mode(false);
```
nit, toggle suggests if it's false, then make it true and vice vesrsa, rather than providing a value to set it
```suggestion
		boosting_with_safe_mode(false);
```

kylezs commented on state-chain/pallets/cf-ingress-egress/src/tests.rs:
```diff
@@ -1794,3 +1794,34 @@ fn ingress_fee_is_withheld_or_scheduled_for_swap() {
+		assert_ok!(IngressEgress::open_channel(
+			&ALICE,
+			eth::Asset::Eth,
+			ChannelAction::LiquidityProvision { lp_account: 0 },
```
Would be another PR - but perhaps it makes sense to block deposit channels for swaps, but not LPs e.g. in a liquidity shortage or something we might want to do that? Something we can think about

dandanlen reviewed (changes requested):
This new safe mode now shadows the `deposits_enabled` safe mode in the swapping pallet. 

I think we should remove the one in the swapping pallet.

msgmaxim commented:
> This new safe mode now shadows the `deposits_enabled` safe mode in the swapping pallet.
> 
> I think we should remove the one in the swapping pallet.

Removed the deposits_enabled field from the swapping pallet now (and the corresponding Error). Also updated the migration code to account for the deleted field. @dandanlen
